HoRNDIS.cpp

```diff
@@ -582,182 +417,297 @@ bool HoRNDIS::createNetworkInterface() {
-	inbuf.comp.action = dataReadComplete;
-	inbuf.comp.parameter = NULL;
-	
-	rtn = fInPipe->Read(inbuf.mdp, &inbuf.comp, NULL);
-	if (rtn != kIOReturnSuccess)
+		fCallbackCount++;
+	}
+
+	// Tell the world that the link is up...
+	if (!setLinkStatus(kIONetworkLinkActive | kIONetworkLinkValid,
+			getCurrentMedium(), getCurrentMedium()->getSpeed())) {
+		LOG(V_ERROR, "Cannot set link status");
+		rtn = kIOReturnError;
 		goto bailout;
+	}
 
-	/* Tell the world that the link is up... */
-	medium = IONetworkMedium::getMediumWithType(fMediumDict, kIOMediumEthernetAuto);
-	setLinkStatus(kIONetworkLinkActive | kIONetworkLinkValid, medium, 480 * 1000000);
-	
-	/* ... and then listen for packets! */
+	// ... and then listen for packets!
 	getOutputQueue()->setCapacity(TRANSMIT_QUEUE_SIZE);
 	getOutputQueue()->start();
 	LOG(V_DEBUG, "txqueue started");
-	
-	/* Tell the other end to start transmitting. */
-	if (!rndisSetPacketFilter(RNDIS_DEFAULT_FILTER))
-		goto bailout;
-	
-	/* Now we can say we're alive. */
+
+	// Now we can say we're alive.
 	fNetifEnabled = true;
-	
-	LOG(V_DEBUG, "done from tid %p", current_thread());
+	LOG(V_NOTE, "completed (thread_id=%lld): tethering interface '%s' "
+		"should be live now", thread_tid(current_thread()), netif->getName());
 	
 	return kIOReturnSuccess;
 	
 bailout:
-	LOG(V_ERROR, "setting up the pipes failed");
-	releaseResources();
+	disableImpl();
 	return rtn;
 }
- 
-IOReturn HoRNDIS::disable(IONetworkInterface * netif) {
-	LOG(V_DEBUG, "disable from tid %p", current_thread());
-	
-	/* Disable the queue (no more outputPacket), and then flush everything in the queue. */
+
+void HoRNDIS::disableNetworkQueue() {
+	// Disable the queue (no more outputPacket),
+	// and then flush everything in the queue.
 	getOutputQueue()->stop();
 	getOutputQueue()->setCapacity(0);
 	getOutputQueue()->flush();
+}
+
+IOReturn HoRNDIS::disable(IONetworkInterface * netif) {
+	LOG(V_DEBUG, "begin for thread_id=%lld", thread_tid(current_thread()));
+	// This function can be called as a consequence of:
+	//  1. USB Disconnect
+	//  2. Some action, while the device is up and running
+	//     (e.g. "ifconfig en6 down").
+	// In the second case, we'll need to do more cleanup:
+	// ask the RNDIS device to stop transmitting, and abort the callbacks.
+	//
+
+	ReentryLocker locker(this, fEnableDisableInProgress);
+	if (locker.isInterrupted()) {
+		LOG(V_ERROR, "Waiting interrupted");
+		return locker.getResult();
+	}
+
+	if (!fNetifEnabled) {
+		LOG(V_DEBUG, "Repeated call (thread_id=%lld)", thread_tid(current_thread()));
+		return kIOReturnSuccess;
+	}
+
+	disableImpl();
+
+	LOG(V_DEBUG, "completed (thread_id=%lld)", thread_tid(current_thread()));
+	return kIOReturnSuccess;
+}
+
+void HoRNDIS::disableImpl() {
+	disableNetworkQueue();
+
+	// Stop the the new transfers. The code below would cancel the pending ones:
+	fReadyToTransfer = false;
+
+	// TODO(mikhailai): Repeated enable/disable (ifconfig up/down)
+	// does not work: investigate!
```

Comment:
Does this cause problems for "reverse tethering" (i.e., internet connection sharing from Mac to RNDIS device)?  I know that is a use case some people care about, especially BeagleBone users.

Reply:
I don't think so. Of course, someone needs to test. I may be getting a PocketBeagle one of these days, so I'll try it out.